contrib/tsan/Make.user.tsan

```diff
@@ -0,0 +1,16 @@
+TOOLCHAIN=$(BUILDROOT)/../toolchain
+BINDIR=$(TOOLCHAIN)/usr/bin
+TOOLDIR=$(TOOLCHAIN)/usr/tools
+
+# use our new toolchain
+USECLANG=1
+override CC=$(BINDIR)/clang
+override CXX=$(TOOLDIR)/clang++
+
+USE_BINARYBUILDER_LLVM=1
```

Comment:
This won't work. You need to build LLVM with tsan instrumentation since it uses locks.

Reply:
Yeah I don't want to build LLVM from source in the Base Julia CI. We'd need to have an "LLVM with TSAN" JLL available from Ygg.

Reply:
> Can we still use a normal-build LLVM with `deadlock:llvm` in the `TSAN_OPTIONS=suppressions=...` configuration? Or there's no way to work around this fundamentally?

It might be possible, but I'd probably recommend against it. I don't know if that can introduce false positives.

Reply:
OK, so maybe it makes sense to close this PR until a JLL is ready.

Or, maybe we can still just run the build process (`-C src`), so that some rudimentary bugs can be caught?

Reply:
> Or, maybe we can still just run the build process (`-C src`), so that some rudimentary bugs can be caught?

I think that's a reasonable place to start.

Reply:
Actually, the latter part is close to what Keno suggested above https://github.com/JuliaLang/julia/pull/42444#issuecomment-931907446

Reply:
> Or, maybe we can still just run the build process (-C src), so that some rudimentary bugs can be caught?

If this part can be done without needing to build LLVM from source, I think it makes sense to go ahead and add it now.

Reply:
Yeah.

Reply:
Yes, no need to instrument LLVM for that.

Reply:
If you can do that in this PR, but leave it as a draft for now, I'm going to promote our current asan job to regular, and then once I've made that change, you can edit this PR so that the "just the part that builds the C runtime with tsan" job goes into the regular pipeline in the same file as the asan job..

Reply:
We can also build LLVM with ccache on CI.

Reply:
> We can also build LLVM with ccache on CI.

When @staticfloat and I last discussed this, our intention was that the Base Julia CI should never be building dependencies from source.

Reply:
I think a `LLVM_tsan_jll` makes the most sense. Then it only needs to be built very occasionally in Ygg, instead of building LLVM from source on every PR that is opened to JuliaLang/julia.

Reply:
I don't really want to support "yet" another version of LLVM in Yggdrasil and slow the update cycle down even more.

Reply:
Hmmm. That's a fair point. Let's see if @staticfloat has any ideas.

At least we can keep the first part, which doesn't require building LLVM from source.

Reply:
What's the source of slowness in yet another LLVM JLL vs ccache? Aren't they equivalent in terms of machine time? Is it that, with the JLL approach, there need to be humans involved in the process waiting for the JLL and updating the URL/version/checksum? If that's the case, it can also potentially be solved by more automation?

I kinda liked the JLL approach since it'd make it easier for more casual users to try sanitizers. OTOH, I guess the ccache approach would be more hackable for core devs since it'd give us a "healthy" CI'ed build script that is easy to flip some LLVM build options. Maybe the upside of the ccache approach wins, since there's no single set of build options that covers all the needs when you debugging things.

Reply:
> Is it that, with the JLL approach, there need to be humans involved in the process waiting for the JLL and updating the URL/version/checksum? If that's the case, it can also potentially be solved by more automation?

Yes, the maintenance of the LLVM JLL is a lot of human time, some of it waiting, some of it making sure that all the version number bits line up. Sheparding things through to the registry. 

I made attempts at trying to automate it, but currently it is pareto optimal in that I find is really annoying, but not annoying enough to descend to arcane wizardy. c.f https://github.com/JuliaPackaging/Yggdrasil/pull/726#issuecomment-617469147